### DISCUSSION TITLE
GLIBC issue

### DISCUSSION THREAD

**Discussion by zero88** [pyinstaller/pyinstaller#5669] [Help] [answered]
Hi community,
I'm new in pyinstaller and not from C lang.
Currently, I'm facing this issue: 
```
[17309] Error loading Python lib '/tmp/_MEIO1NXeM/libpython3.8.so.1.0': dlopen: /lib/arm-linux-gnueabihf/libc.so.6: version `GLIBC_2.25' not found (required by /tmp/_MEIO1NXeM/libpython3.8.so.1.0)
```

Ya, I knew almost suggestion to update glibc version or i should re-compiled as per server version. But it is not easy in my case due to:
- my target environment is quite dynamic (IoT device, on production then it is quite hard to upgrade glibc and OS version)
- my build environment is depended on virtual environment of github action

So, I want to ask if any solution to use `glibc` statically in my binary instead of using environment glibc. 
Something like, I can download a specified `glibc.so` into lib folder, then instruct pyinstaller compile and link to it from my code.

Thanks

**Comment by zero88** [comment]
Does `LD_LIBRARY_PATH` help?

**Comment by bwoodsend** [comment, marked as answer]
@rokm will know more about this than me but generally copying system libraries like `glibc` rarely works. You're better off finding a Python built with a lower `glibc` requirement which in practice means building on an old version of Linux or an old docker image. [cdrx/docker-pyinstaller](https://github.com/cdrx/docker-pyinstaller) is a bit out of date but is otherwise ideal for this.

**Reply by johnthagen** [reply]
Here's another (recently updated) take on building Pyinstaller binaries in Docker: https://github.com/fyde/pyinstaller

**Comment by rokm** [comment]
I don't think overriding the `libc.so` via `LD_LIBRARY_PATH` or `LD_PRELOAD` will work (see [this](https://stackoverflow.com/questions/847179/multiple-glibc-libraries-on-a-single-host/851229#851229)).

You'll either need to freeze on a machine whose `glibc` version is as low as the lowest `glibc` version you expect to encounter on your target systems, or you could try post-processing your `onefile` executable with `staticx` (but if you do, make sure you use `develop` PyInstaller version, and that you rebuild the bootloader, so that #5511 is in effect - otherwise, you'll likely hit #5330).

**Comment by zero88** [comment]
Thank you for your response, appreciate it.

I already resolved my situation by docker.
Currently, I'm building a multiple arch application on docker by official python image, then I'm able to downgrade OS version to make GLIBC compatible, but downgrade python version also. Luckily, my library is standalone cli without too much dependencies, then it is trivial.

So, if anyone encounter this issue, use docker is best solution. however, with new python in old OS, for example: 3.9 on stretch need to build your docker file manually.

**Comment by johnthagen** [comment]
Another strategy for bundling a newer version of Python that targets an older version of GLIBC is to use the [Ubuntu Deadsnakes PPA](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa) to install a newer Python on an older Ubuntu.

An example `Dockerfile` (Python 3.9 on Ubuntu 18.04):

```Dockerfile
FROM ubuntu:18.04

RUN apt-get update && apt-get install -y software-properties-common binutils
RUN add-apt-repository ppa:deadsnakes/ppa
RUN apt-get update && apt-get install -y python3.9 python3.9-dev python3.9-venv
RUN python3.9 -m venv /venv --upgrade-deps

COPY main.py .

RUN export PATH=/venv/bin:$PATH; python -m pip install pyinstaller && \
  pyinstaller --onefile main.py
```

**Reply by bwoodsend** [reply]
That's quite a nice one. Much cleaner than CentOS7 that PyPA traditionally use. And certainly beats monkeying around with `pyenv`.
